### PR TITLE
client: reset backoff to 0 after a connection is established

### PR DESCRIPTION
**Pull request by menghanl**
#2663
#2636

### CODE REVIEW

File: clientconn.go [hunk 1]
```diff
@@ -1032,6 +1032,7 @@ func (ac *addrConn) resetTransport() {
 				continue
 			}
 
+			backoffFor = 0
```
Comment (kazegusuri) [inline clientconn.go]: It works fine for us to reconnect immediately after connection closed.
But I think this also resets backoff duration when reconnecting failed repeated in the server down, which resulted in trying to reconnect aggressively. 

So the backoff duration should be reset only in first retry like this?
```
if ac.backoffIdx == 0 {
	backoffFor = 0
}
```

**Comment by dfawley** [comment]
@kazegusuri 

If the connection fails to be created, this code will be skipped.

Did you apply this patch to master@HEAD or another branch/commit?

Are you setting `GRPC_GO_REQUIRE_HANDSHAKE=off` in your environment?  If so, please try not setting it.

File: clientconn.go (hunk 1 above)
Comment (dfawley) [inline clientconn.go]: Actually, this should also reset `backoffIdx = 0`.

Comment (menghanl) [reply clientconn.go]: I have cleanup changes in another branch, was trying to keep this change simple so we can cherrypick it to previous releases.

Comment (menghanl) [reply clientconn.go]: If reconnecting failed, we will `continue` above, and won't reach this.
This only happens when we successfully create a connection, and we want to skip any backoff if this connection breaks.

Comment (kazegusuri) [reply clientconn.go]: sorry it's my bad. I tried again with this branch and it does retry to connect with backoff correctly.

**Pull request merged**